blockchain/chain_sync/src/sync.rs

```diff
@@ -542,8 +797,339 @@ where
+        let rpc_response = RPCResponse::BlockSync(BlockSyncResponse {
+            chain: vec![
+                construct_tipset_bundle(3),
+                construct_tipset_bundle(2),
+                construct_tipset_bundle(1),
+            ],
+            status: 0,
+            message: "message".to_owned(),
+        });
+        let c_response = rpc_response.clone();
+
+        cs.net_handler.spawn(Arc::clone(&cs.peer_manager));
+
+        // ensure response is on channel prior to spawning test thread
+        task::block_on(async {
+            event_sender
+                .send(NetworkEvent::RPCResponse {
+                    req_id: 0,
+                    response: c_response.clone(),
+                })
+                .await;
+        });
+
+        // params for sync_headers_reverse
+        let source = PeerId::random();
+        let head = construct_tipset(4);
+        let to = construct_tipset(1);
+
+        task::block_on(async move {
+            cs.peer_manager.add_peer(source.clone()).await;
+            assert_eq!(cs.peer_manager.len().await, 1);
+
+            let return_set = cs.sync_headers_reverse(head, &to).await;
+            assert_eq!(return_set.unwrap().len(), 4);
+        });
+    }
+
+    #[test]
+    fn inform_new_head_given_fts_test() {
+        let fts = construct_full_tipset();
+        let source = PeerId::random();
+
+        let db = MemoryDB::default();
+        let mut cs = chain_syncer_setup(Arc::new(db));
+
+        task::spawn(async move {
+            cs.inform_new_head(&source, &fts)
+                .await
+                .expect("error with informing of new head");
+        });
+    }
+
+    #[test]
+    fn set_peer_head_and_select_target_test() {
+        let ts = construct_tipset(1);
+        let source = PeerId::random();
+
+        let db = MemoryDB::default();
+        let mut cs = chain_syncer_setup(Arc::new(db));
+
+        task::spawn(async move {
+            cs.set_peer_head(&source, ts.clone())
+                .await
+                .expect("error setting peer head");
+            for (peer_id, tip) in cs.peer_manager.peer_heads.read().await.iter() {
+                assert_eq!(peer_id, &source);
+                assert_eq!(tip, &ts.clone());
+            }
+            assert_eq!(cs.select_sync_target().await.unwrap(), ts);
+        });
+    }
+
+    #[test]
+    fn schedule_tipset_given_tipset_test() {
+        let ts = construct_tipset(1);
+        let ts_copy = construct_tipset(1);
+
+        let db = MemoryDB::default();
+        let mut cs = chain_syncer_setup(Arc::new(db));
+
+        task::spawn(async move {
+            cs.schedule_tipset(ts.clone())
+                .await
+                .expect("error scheduling tipsets");
+            assert_eq!(cs.next_sync_target._tipsets(), &[ts]);
+
+            cs.schedule_tipset(ts_copy)
+                .await
+                .expect("error scheduling tipsets");
+            assert_eq!(cs.next_sync_target._tipsets().len(), 2);
```

Comment:
Yes, that is correct. I will update the test to make them different tipsets. The reason this did not fail is because the logic for identifying and handling related tipsets is a TODO.